feature/bookmarks/src/test/kotlin/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksViewModelTest.kt

```diff
@@ -88,6 +90,54 @@ class BookmarksViewModelTest {
         val item = viewModel.feedUiState.value
         assertIs<Success>(item)
         assertEquals(item.feed.size, 0)
+        assertTrue(viewModel.shouldDisplayUndoBookmark)
+
+        collectJob.cancel()
+    }
+
+    @Test
+    fun whenResourceViewed_setResourcesViewed() = runTest {
+        val collectJob = launch(UnconfinedTestDispatcher()) { viewModel.feedUiState.collect() }
+
+        // Give
+        newsRepository.sendNewsResources(newsResourcesTestData)
+        userDataRepository.setNewsResourceBookmarked(newsResourcesTestData[0].id, true)
+        val itemBeforeViewed = viewModel.feedUiState.value
+        check(itemBeforeViewed is Success)
+        check(!itemBeforeViewed.feed.first().hasBeenViewed)
+
+        // When
+        viewModel.setNewsResourceViewed(newsResourcesTestData[0].id, true)
+
+        // Then
+        val item = viewModel.feedUiState.value
+        assertIs<Success>(item)
+        assertTrue(item.feed.first().hasBeenViewed)
+
+        collectJob.cancel()
+    }
+
+    @Test
+    fun whenUndoBookmarkRemoval_thenBookmarkIsRestored() = runTest {
+        val collectJob = launch(UnconfinedTestDispatcher()) { viewModel.feedUiState.collect() }
+
+        // Give
+        newsRepository.sendNewsResources(newsResourcesTestData)
+        userDataRepository.setNewsResourceBookmarked(newsResourcesTestData[0].id, true)
+        viewModel.removeFromSavedResources(newsResourcesTestData[0].id)
+        check(viewModel.shouldDisplayUndoBookmark)
+        val itemBeforeUndo = viewModel.feedUiState.value
+        check(itemBeforeUndo is Success)
+        check(itemBeforeUndo.feed.isEmpty())
```

Comment:
```suggestion
        assertTrue(viewModel.shouldDisplayUndoBookmark)
        val itemBeforeUndo = viewModel.feedUiState.value
        assertIs<Success>(itemBeforeUndo)
        assertEquals(0, itemBeforeUndo.feed.size)
```

Also, I'm not sure these assertions in the "given" block are really useful.

Reply:
@SimonMarquis 
Thank you for your review!

The reason I implement the assertion in Given block is simply readability.
Without the given block assertion, we need to follow newsResourcesTestData  and the process of BookmarksViewModel to understand the scenario.

However, this may be too redundant.